xtra-libp2p-offer/src/maker.rs

```diff
@@ -0,0 +1,112 @@
+use crate::protocol;
+use crate::PROTOCOL_NAME;
+use async_trait::async_trait;
+use model::MakerOffers;
+use std::collections::HashSet;
+use std::time::Duration;
+use tokio_tasks::Tasks;
+use xtra::spawn::TokioGlobalSpawnExt;
+use xtra::Actor as _;
+use xtra::Context;
+use xtra_libp2p::libp2p::PeerId;
+use xtra_libp2p::Endpoint;
+use xtra_libp2p::GetConnectionStats;
+use xtra_libp2p::OpenSubstream;
+use xtra_productivity::xtra_productivity;
+use xtras::spawner;
+use xtras::spawner::SpawnFallible;
+use xtras::SendAsyncSafe;
+use xtras::SendInterval;
+
+const UPDATE_CONNECTED_PEERS_INTERVAL: Duration = Duration::from_secs(5);
+
+pub struct Actor {
+    endpoint: xtra::Address<Endpoint>,
+    connected_peers: HashSet<PeerId>,
+    spawner: xtra::Address<spawner::Actor>,
+    tasks: Tasks,
+}
+
+impl Actor {
+    pub fn new(endpoint: xtra::Address<Endpoint>) -> Self {
+        let spawner = spawner::Actor::new().create(None).spawn_global();
+
+        Self {
+            endpoint,
+            connected_peers: HashSet::default(),
+            spawner,
+            tasks: Tasks::default(),
+        }
+    }
+}
+
+#[xtra_productivity]
+impl Actor {
+    async fn handle(&mut self, msg: NewOffers) {
+        for peer in self.connected_peers.iter().copied() {
+            let endpoint = self.endpoint.clone();
+            let offers = msg.0.clone();
+
+            let task = async move {
+                tracing::trace!(%peer, "Sending offers");
+
+                let stream = endpoint
+                    .send(OpenSubstream::single_protocol(peer, PROTOCOL_NAME))
+                    .await??;
+
+                protocol::send(stream, offers).await?;
+
+                anyhow::Ok(())
+            };
+
+            let err_handler =
+                move |e| async move { tracing::debug!(%peer, "Failed to send offers: {e:#}") };
+
+            if let Err(e) = self
+                .spawner
+                .send_async_safe(SpawnFallible::new(task, err_handler))
+                .await
+            {
+                tracing::warn!("Failed to spawn task to send offers: {e:#}");
+            };
+        }
+    }
+
+    async fn handle(&mut self, _: UpdateConnectedPeers) {
```

Comment:
Since we are keeping the pattern of pushing the offers from maker to taker https://github.com/itchysats/itchysats/pull/2141 is actually relevant for this PR.
With https://github.com/itchysats/itchysats/pull/2141 we will be able to trigger an `offer` update upon a new connection (and not just when upon `NewOffers` message). This will make the UX slightly better, because  `NewOffers` is not guaranteed to happen in a fixed (short) interval. It might be worth rebasing this work on top of https://github.com/itchysats/itchysats/pull/2141

Reply:
exactly, it would be great to use `EndpointMonitor` to stop polling the Endpoint manually (less chance for message slowdown)

Reply:
I agree! I just wanted to get something working to open the PR.

PR #2141 is pretty much necessary to merge this, otherwise this introduces the regression that you're hinting at: a newly connected taker may not hear about the latest offers for a while.

This can be seen in the actor tests failing, since they must have been written with the assumption that a new taker will know the latest offers as soon as they connect.

In conclusion, I will wait for #2141 to merge before fixing CI on this PR.